healthmanager/subscription_manager_test.go

```diff
@@ -0,0 +1,117 @@
+package healthmanager
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+)
+
+func TestSubscriptionManager(t *testing.T) {
+	// Create a new SubscriptionManager
+	sm := NewSubscriptionManager()
+
+	// Create a context for cancellation
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Create a WaitGroup to wait for all goroutines to finish
+	var wg sync.WaitGroup
+
+	// Test adding subscribers and receiving notifications
+	subscriberCount := 5
+	notificationReceived := make([]bool, subscriberCount)
+
+	for i := 0; i < subscriberCount; i++ {
+		ch := sm.Subscribe()
+		idx := i // Copy the value of i to use inside the goroutine
+
+		wg.Add(1)
+		go func(ch chan struct{}, idx int) {
+			defer wg.Done()
+			select {
+			case <-ch:
+				notificationReceived[idx] = true
+			case <-time.After(1 * time.Second):
+				// Timeout waiting for notification
+			}
+		}(ch, idx)
+	}
+
+	// Emit a notification to all subscribers
+	sm.Emit(ctx)
+
+	// Wait for all goroutines to finish
+	wg.Wait()
+
+	// Verify that all subscribers received the notification
+	for i, received := range notificationReceived {
+		if !received {
+			t.Errorf("Subscriber %d did not receive notification", i)
+		}
+	}
+
+	// Test unsubscribing a subscriber
+	ch := sm.Subscribe()
+
+	// Unsubscribe the subscriber
+	sm.Unsubscribe(ch)
+
+	// Attempt to send a notification
+	sm.Emit(ctx)
+
+	select {
+	case _, ok := <-ch:
+		if ok {
+			t.Errorf("Unsubscribed channel received notification")
+		}
+	case <-time.After(100 * time.Millisecond):
+		// Expected behavior: no notification should be received
```

Comment:
Not quite good thing really: [**Prevent flaky tests:** don’t wait for absence of something](https://www.notion.so/Prevent-flaky-tests-2045334180cb4eabbe2151f7005e16ce?pvs=4#e77e978d5a694b2dac7552ef68238913)

Not sure what's the best approach in this case though. But perhaps just checking that the channel doesn't exist in the `SubscriptionManager.subscription` is enough. You have access to the private variable here.

Reply:
fixed
Thanks for pointing out this page!